internal/provider/services/gateway/gateway.go

```diff
@@ -0,0 +1,108 @@
+)
+
+type TerrariumGrpcGateway struct {
+	terrarium.UnimplementedPublisherServer
+	versionManagerClient services.VersionManagerClient
+}
+
+func New(versionManagerClient services.VersionManagerClient) *TerrariumGrpcGateway {
+	return &TerrariumGrpcGateway{
+		versionManagerClient: versionManagerClient,
+	}
+}
+
+// RegisterWithServer registers TerrariumGrpcGateway with grpc server
+func (gw *TerrariumGrpcGateway) RegisterWithServer(grpcServer grpc.ServiceRegistrar) error {
+	terrarium.RegisterPublisherServer(grpcServer, gw)
+	return nil
+}
+
+// Register new provider with Registrar service
+func (gw *TerrariumGrpcGateway) Register(ctx context.Context, request *terrarium.RegisterProviderRequest) (*terrarium.Response, error) {
+	return gw.RegisterWithClient(ctx, request, gw.versionManagerClient)
+}
+
+// RegisterWithClient calls Register on version manager client
+func (gw *TerrariumGrpcGateway) RegisterWithClient(ctx context.Context, request *terrarium.RegisterProviderRequest, client services.VersionManagerClient) (*terrarium.Response, error) {
+	span := trace.SpanFromContext(ctx)
+	span.SetAttributes(
+		attribute.String("provider.name", request.GetName()),
+		attribute.String("provider.version", request.GetVersion()),
+	)
+
+	if res, delegateError := client.Register(ctx, request); delegateError != nil {
+		log.Printf("Failed: %v", delegateError)
+		span.RecordError(delegateError)
+		return nil, delegateError
+	} else {
+		log.Println("Done <= Registered")
+		span.AddEvent("Successful call to Register on Version Manager client.")
+		return res, nil
+	}
+}
+
+func (gw *TerrariumGrpcGateway) EndProvider(ctx context.Context, request *terrarium.EndProviderRequest) (*terrarium.Response, error) {
+	return gw.EndProviderWithClient(ctx, request, gw.versionManagerClient)
+}
+
+// EndVersionWithClient calls AbortProvider/AbortProviderVersion/PublishVersion on Version Manager client
+func (gw *TerrariumGrpcGateway) EndProviderWithClient(ctx context.Context, request *terrarium.EndProviderRequest, client services.VersionManagerClient) (*terrarium.Response, error) {
+
+	terminateProviderVersion := services.TerminateVersionRequest{
+		Provider: request.GetProvider(),
+	}
+
+	publishRequest := services.TerminateVersionRequest{
+		Provider: request.GetProvider(),
+	}
+
+	span := trace.SpanFromContext(ctx)
+
+	if request.GetAction() == terrarium.EndProviderRequest_DISCARD_VERSION {
+		log.Println("Abort Provider Version => Version Manager")
+		if res, delegateError := client.AbortProviderVersion(ctx, &terminateProviderVersion); delegateError != nil {
+			log.Printf("Failed: %v", delegateError)
+			span.RecordError(delegateError)
+			return nil, delegateError
+		} else {
+			log.Println("Done <= Version Manager")
+			span.AddEvent("Successfully aborted provider version", trace.WithAttributes(attribute.String("Provider Name", request.Provider.GetName()), attribute.String("Provider Version", request.Provider.GetVersion())))
```

Comment:
Don't need this if you specify the span attributes. Also this event doesn't really add any value - the usual case would be for the operation to succeed, therefore the more interesting case is if something goes wrong where you're recording the error.

Reply:
removed span events.